Update BigQuerySinkMetrics for StreamingInserts.

JayajP opened the pull request:
Update the following metrics after we insert a batch of rows using BigQuery's InsertAll RPC.

### 1. `RowsAppendedCount` Counter
Tracks the status of BigQuery Rows after the batch of InsertAll RPCs is completed.

Metric has labels the following labels:
- `RowStatus`: Status of the BigQuery rows after , one of `SUCCESSFUL`, `RETRIED`, `FAILED`
-  `RpcStatus`: Rpc Status
- `TableId`: 'datasets/{ }/tables/{ }' that the rows are sent to.

Rows are updated based on the following criteria.
- Rows that are successfully inserted will have their `RowStatus` set to successful.
- Rows that have to be retried because the RPC failed will have their `RowStatus` set to `RETRIED` and `RpcStatus` set to the failed RPC's status.
- Rows might fail to be inserted even when the `InsertAll` RPC succeeds. These rows will have their `RowStatus` set to `RETRIED` and `RpcStatus` set to `INTERNAL`.
- Rows that fail to be inserted will `RowStatus` set to `RETRIED` and `RpcStatus` set to `INTERNAL`.

### 2. `RpcRequests` Counter 
Tracks the Rpc Status of the InsertAll Rpc.

Metric has labels the following labels:
- `Method`: BigQuery sink method. Will be `STREAMING_INSERTS` for these metrics.
-  `RpcStatus`: Rpc Status
- `TableId`: 'datasets/{ }/tables/{ }' that the rows are sent to.

### 3. `RpcLatency` Histogram:
Tracks the Rpc Latency of the InsertAll Rpc. 

Metric has labels the following labels:
- `Method`: BigQuery sink method. Will be `STREAMING_INSERTS` for these metrics.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

m-trieu commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySinkMetrics.java:
```diff
@@ -326,6 +337,83 @@ public static void reportFailedRPCMetrics(
+
+    String shortTableId =
+        String.format("datasets/%s/tables/%s", tableRef.getDatasetId(), tableRef.getTableId());
+    Map<String, Integer> rpcRequetsRpcStatusMap = new HashMap<>();
```
see comment above about atomicity for the concurrent collections.
They allow concurrent writes to the data structure (i.e you could be iterating/reading here, while updates are happening to the data structure without ConcurrentModificationException, it doesn't guarantee that you will see every update while iterating -`Iterators are weakly consistent, returning elements reflecting the state of the queue at some point at or since the creation of the iterator. They do not throw [ConcurrentModificationException](https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/util/ConcurrentModificationException.html), and may proceed concurrently with other operations. Elements contained in the queue since the creation of the iterator will be returned exactly once.` from javadoc)

if you want (not sure if you do), to tie the operations together (the iterating over `results.rpcStatus` and `results.retriedRowsByStatus`) you will need to wrap the iteration in a lock like :

```
synchronized (someObject) // if instance method in StreamingInsertsResults, can use synchronized(this)
    for (String status : results.rpcStatus) {
      Integer currentVal = rpcRequetsRpcStatusMap.getOrDefault(status, 0);
      rpcRequetsRpcStatusMap.put(status, currentVal + 1);
    }

    for (KV<String, Integer> retryCountByStatus : results.retriedRowsByStatus) {
      Integer currentVal = retriedRowsRpcStatusMap.getOrDefault(retryCountByStatus.getKey(), 0);
      retriedRowsRpcStatusMap.put(
          retryCountByStatus.getKey(), currentVal + retryCountByStatus.getValue());
    }
}
```

depending on the behavior you want, if you don't need to tie these 2 operations together than its ok to leave as is.

JayajP replied:
Don't need synchronization across member variables

m-trieu commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySinkMetrics.java:
```diff
@@ -326,6 +337,83 @@ public static void reportFailedRPCMetrics(
+   * Results of a batch of InsertAll RPCs. Used to update metrics with the {@code
+   * updateStreamingInsertsMetrics} method. Member variables should be thread-safe.
+   */
+  public static class StreamingInsertsResults {
```
if this is a data class, can you use AutoValue? 
https://g3doc.corp.google.com/third_party/java_src/auto/value/g3doc/index.md?cl=head

can also annotate with `@ThreadSafe`

also the members are threadsafe individually, but if you need atomic operations across multiple members that will still require a lock (`synchronized (someObject)`)

JayajP replied:
Added a more nuanced comment about the thread safety guarantees of this class.

m-trieu commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java:
```diff
@@ -1258,6 +1278,10 @@ <T> long insertAll(
           }
         }
       }
+      streamingInsertsResults.failedRowsCount.getAndAdd(allErrors.size());
```
see comment about making instance method
maybe this one could be something like:
```
@AutoValue
public abstract static StreamingInsertResults {
...
void updateRowCount(int failedRowCount, int successRowCount) {
    failedRowsCount().getAndAdd(failedRowCount);
    successfulRowsCount().set(successRowCount - failedRowsCount.get())
}
...
}
```

JayajP replied:
Done.

m-trieu commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySinkMetrics.java:
```diff
@@ -326,6 +337,83 @@ public static void reportFailedRPCMetrics(
+   * @param results Result of StreamingInsert RPC calls
+   * @param tableRef BigQuery table that was written to.
+   */
+  public static void updateStreamingInsertsMetrics(
```
when would the inputs be null?
mark results and tableRef as `@Nullable` like 
`(@Nullable StreamingInsertsResults results, @Nullable TableReference tableRef)`

m-trieu replied:
ah gotcha thats unfortunate sgtm

m-trieu commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySinkMetrics.java:
```diff
@@ -326,6 +337,83 @@ public static void reportFailedRPCMetrics(
+   *
+   * @param results Result of StreamingInsert RPC calls
+   * @param tableRef BigQuery table that was written to.
+   */
```
how about making this an instance method w/in `StreamingInsertsResults`?

can be like 
```
public void updateMetrics(@Nullable TableReference tableRef) {

...//do operations on members instead of passed in StreamingInsertsResults.

}
```

this can be done even with AutoValue

JayajP replied:
done.

JayajP commented:
Thanks for the review Martin, switched `StreamingInsertsResults` to an AutoValue class, made the instance-variables private and added some update methods to update the instance-variables.

m-trieu commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySinkMetrics.java:
```diff
@@ -326,6 +338,127 @@ public static void reportFailedRPCMetrics(
+        new ConcurrentLinkedQueue<>();
+    private final AtomicInteger successfulRowsCount = new AtomicInteger();
+    private final AtomicInteger failedRowsCount = new AtomicInteger();
+    private final AtomicBoolean isWritable = new AtomicBoolean(true);
```
close!

for AutoValue magic these will need to be defined as abstract getters() like:

```
abstract static class StreamingInsertsResults {
    abstract ConcurrentLinkedQueue<java.time.Duration> rpcLatencies();
    abstract  ConcurrentLinkedQueue<String> rpcStatus();
    .....etc.
}
```

then you can expose whatever in you static create method and pass the default values of the members to `new AutoValue_BigQuerySinkMetric_StreamingInsertsResults(new ConcurrentLinkedQueue<>() .....);

rest looks good

m-trieu reviewed (approved):
Besides changing to match AutoValue syntax LGTM

m-trieu reviewed (approved):
LGTM
